spotify_tensorflow/dictionary_endpoint.py

```diff
@@ -0,0 +1,124 @@
+        :param batch_size: Size of each batches
+        :param unpack_multispec: Returns an array of DataFrames, order is the same as the one used
+        when building the MultiFeatureSpec in Featran
+        :param feature_mapping_fn: Override the TF record reading function
+        :return: A Python Generator, yielding batches of data in a Dictonary
+        """
+        training_it, context = Datasets.mk_iter(dataset_path, feature_mapping_fn=feature_mapping_fn)
+        for batch in cls.__FeatureGenerator(training_it, batch_size, context):
+            yield batch
+
+    class __FeatureGenerator(object):
+        def __init__(self, training_it, batch_size, context):
+            self.batch_size = batch_size
+            self.batch_iter = training_it.get_next()
+            self.context = context
+            self.buff = None
+
+        def __iter__(self):
+            logger.info("Starting TF Session...")
+            with tf.Session() as sess:
+                logger.info("Reading TFRecords...")
+                while True:
+                    try:
+                        yield self._get_batch(sess)
+                    except tf.errors.OutOfRangeError:
+                        logger.info("End of dataset.")
+                        break
+            yield self.buff
+
+        def _get_buff_size(self):
+            if self.buff is None:
+                return 0
+            else:
+                return len(self.buff[list(self.buff.keys())[0]])
+
+        def _append(self, v1, v2):
+            if type(v1) is np.ndarray:
+                if(v1.ndim == 1):
+                    return np.append(v1, v2)
+                elif(v1.ndim == 2):
+                    return np.hstack([v1, v2])
+                else:
+                    raise ValueError("Only 1 or 2 dimensional features are supported")
+            else:
+                return v1.append(v2)
+
+        def _get_batch(self, sess):
+            if self.buff is None:
+                self.buff = OrderedDict()
+                first_result = sess.run(self.batch_iter)
+                for k in self.context.features.keys():
+                    self.buff[k] = first_result[k]
+            while self._get_buff_size() < self.batch_size:
+                t = timeit.default_timer()
+                current_batch = sess.run(self.batch_iter)
+                for k in self.context.features.keys():
+                    self.buff[k] = self._append(self.buff[k], current_batch[k])
+                logger.info("Fetched %d / %s records (%4d TFExamples/s)" % (
+                    self._get_buff_size(),
+                    str(self.batch_size) if self.batch_size < sys.maxsize else "?",
+                    FLAGS.batch_size / (timeit.default_timer() - t)))
+            ret = OrderedDict()
+            for k in list(self.context.features.keys()):
+                ret[k] = self.buff[k][:self.batch_size]
+                self.buff[k] = self.buff[k][self.batch_size:]
```

Comment:
What is this line for?

Reply:
This is to segment the currently buffered data into the values to return and to keep the rest stored on the buff object.  Because each read can be a different size than the batch size.